feat(core): support TypeScript 4.4

crisbeto opened the pull request:
Adds support for TypeScript 4.4. High-level overview of the changes made in this PR:

* Bumps the various packages to `typescript@4.4.2` and `tslib@2.3.0`.
* The `useUnknownInCatchVariables` compiler option has been disabled so that we don't have to cast error objects explicitly everywhere.
* TS now passes in a third argument to the `__spreadArray` call inside child class constructors. I had to update a couple of places in the runtime and ngcc to be able to pick up the calls correctly.
* TS now generates code like `(0, foo)(arg1, arg2)` for imported function calls. I had to update a few of our tests to account for it. See https://github.com/microsoft/TypeScript/pull/44624.
* Our `ngtsc` test setup calls the private `matchFiles` function from TS. I had to update our usage, because a new parameter was added.
* There was one place where we were setting the readonly `hasTrailingComma` property. I updated the usage to pass in the value when constructing the object instead.
* Some browser types were updated which meant that I had to resolve some trivial type errors.
* The downlevel decorators tranform was running into an issue where the Closure synthetic comments were being emitted twice. I've worked around it by recreating the class declaration node instead of cloning it.

JoostK commented:
Just a heads up from the ngcc perspective: it has some integration tests that use TypeScript to compile TS sources to JS as that was generally a good way to know that ngcc is capable of processing the output from TS, but now that we're transitioning to Ivy partial compilation and compiling libs with View Engine will not be supported in v13 there isn't necessarily a need for ngcc to process libraries that have been compiled using TS 4.4. We may want to revise the tests to stick to TS 4.3, or maybe there's an alternative approach.

crisbeto commented on packages/compiler-cli/test/transformers/downlevel_decorators_transform_spec.ts:
```diff
@@ -182,7 +182,7 @@ describe('downlevel decorator transform', () => {
     expect(diagnostics.length).toBe(0);
     expect(output).toContain(dedent`
-      MyClass = tslib_1.__decorate([
+      MyClass = (0, tslib_1.__decorate)([
```
These look weird, but they're intentional. See https://github.com/microsoft/TypeScript/pull/44624.

IgorMinar commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/circle-ci-api.ts:
```diff
@@ -22,28 +22,28 @@ export interface BuildInfo {
-    private githubOrg: string,
-    private githubRepo: string,
-    private circleCiToken: string,
+      private githubOrg: string,
```
revert?

atscott reviewed (approved):
reviewed-for: fw-router

IgorMinar commented on packages/compiler-cli/src/typescript_support.ts:
```diff
@@ -25,7 +25,7 @@ const MIN_TS_VERSION = '4.2.3';
  * Note: this check is disabled in g3, search for
```
please update MIN_TS_VERSION to 4.4.2

IgorMinar reviewed (approved):
Thanks @crisbeto!  This looks good, with a few remarks:
- I'm not too happy with all the `any`s all over the place, but I understand why you went down that route. I wish TS was a bit more accommodating here and allowed us to specify the type of the argument in the `catch` signature.
- the `addSyntheticLeadingComment` duplication might actually be a problem - please sync up with @alxhub who can verify. JSCompiler can be get upset when it sees duplicate or unknown comments.
- please drop TS 4.2 and 4.3 support, but keep the integration test for 4.3 to ensure that we are still g3 compatible. g3 is moving to 4.4 as we speak, so we need the test only for a short period of time.


thanks again!

IgorMinar reviewed (approved):
Reviewed-for: global-approvers

alan-agius4 commented:
//@crisbeto and @IgorMinar 

Since we are not verifying whether the error is an `instanceOf` an `Error`, `string` or something else, and many cases we are assuming it is an instance of an Error. I don't see benefit of using the `useUnknownInCatchVariables` compiler TSC option which is now the default in v4.4 and in fact is the root cause of the newly added castings.

Therefore, why not disable this in option in the first place and retain the old behaviour? Or this is left on, so that in future written catch clause we will enforce checking the type of an error similar the below?

```ts
try {
} catch (err) {
  if (err instanceof Error) {
    console.error(error.message);
  } else {
    console.error(err); 
  }
}
```

crisbeto commented:
The feedback has been addressed, but I have a couple of notes:

* I set the `MIN_TS_VERSION` to 4.3.5 for now, because the Components repo is still on it and the CI check will fail. We can bump it once Components has been updated.
* I'm marking the PR as blocked for now, until we can figure out what to do about https://github.com/microsoft/TypeScript/issues/45696 which affects our code. I see the following options going forward:
* 1. Keep the duplicated comments as they are, assuming nothing break in g3.
* 2. Use the workaround I have in this PR which uses `createClassDeclaration` instead of `updateClassDeclaration`.
* 3. Wait for the TS issue to be resolved.

IgorMinar commented:
@alan-agius4 good point. I asked the internal team about the plan for g3.

I think you bring up an interesting question @crisbeto what's your take?

I'm not a fan of all the casts especially casts to any. So I'd love to consider all the options.

crisbeto commented:
Disabling `useUnknownInCatchVariables` sounds good to me, but we'll have to check if it'll have an impact in g3. I'm not a fan of the casts to `any` either, but I'm hesitant to include more `instanceof` checks, because they'll add some bytes to the bundle size.

atscott commented:
[presubmit link](https://fusion2.corp.google.com/presubmit/tap/395529752/targets?utm_source=tap-redirect)

h-joo commented:
Sorry for the late reply. With respect to google codebase, we don't have immediate plans to turn on `--useUnknownInCatchVariables` flag in google.

mprobst commented:
Regarding the duplicated comments issue, I have a PR at https://github.com/microsoft/TypeScript/pull/45818 to fix that (though it might need some more work).

AndrewKushnir commented:
[Presubmit](http://test/OCL:396625176:BASE:396625226:1631639926095:1a6f39af).

crisbeto commented:
As discussed, I've reverted the error objects casts and disabled the `useUnknownInCatchVariables` compiler flag. I've also rebased and fixed some errors coming in from master.

crisbeto commented:
PR to update the `api-extractor` to TS 4.4 so we don't have to work around it: https://github.com/microsoft/rushstack/pull/2912

AndrewKushnir commented:
Quick update: global presubmit indicated some failures and I've narrowed them down to this change in the PR (the `packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts` file):

```
      const newClassDecl = ts.createClassDeclaration(
          decoratorsToKeep.size ? Array.from(decoratorsToKeep) : undefined, classDecl.modifiers,
          classDecl.name, classDecl.typeParameters, classDecl.heritageClauses, members);
      return ts.setTextRange(newClassDecl, classDecl);
```

If I revert back that change, I see no issues. The error that is being thrown:

```
some-token.closure.js: ERROR - [JSC_TOO_MANY_TEMPLATE_PARAMS] Too many template parameters
Found 1, required at most 0
  17|  * @type {!tsickle_injection_token_3.InjectionToken<!tsickle_some_token_2.SomeToken>}
```

The original code looks like this (replaced real names with test ones):

```
/**
 * Some token.
 */
export const SOME_MAP = new InjectionToken<SomeToken>('some-token');
```

Genrated output is:

```
/**
 * Some token.
 * @type {!tsickle_injection_token_3.InjectionToken<!tsickle_some_token_2.SomeToken>}
 */
exports.SOME_TOKEN = new core_1.InjectionToken('some-token');
```

One observation is that the generated code with and without the changes from this PR is the same, but the error is there for some reasons. I suspect it might be related to the fact that we've used to update existing class and now we generate a new one, thus Tsickle gets confused and retains the reference to the old class.

cc @alxhub

crisbeto commented:
Thank you for investigating @AndrewKushnir. As mentioned above, the code you linked is a workaround for https://github.com/microsoft/TypeScript/issues/45696, because we were under the assumption that the duplicated comments would cause issues in g3. I've pushed a commit that reverts the workaround.

mprobst commented:
It might be better to wait for my PR to land?

Re @AndrewKushnir's issue, in general tsickle needs to know the "original node" of whatever it is transforming, so that it can learn its type and then emit the signature correctly. If you just create a new class node or similar, it won't have its type attached (or checked), and thus type emit may be wrong.

h-joo commented:
Would you think it's a good idea(or possible) to make this changes in without bumping the TS version to make google's codebase move forward, but not affecting other users?

crisbeto committed:
feat(core): support TypeScript 4.4

Adds support for TypeScript 4.4. High-level overview of the changes made in this PR:

* Bumps the various packages to `typescript@4.4.2` and `tslib@2.3.0`.
* The `useUnknownInCatchVariables` compiler option has been disabled so that we don't have to cast error objects explicitly everywhere.
* TS now passes in a third argument to the `__spreadArray` call inside child class constructors. I had to update a couple of places in the runtime and ngcc to be able to pick up the calls correctly.
* TS now generates code like `(0, foo)(arg1, arg2)` for imported function calls. I had to update a few of our tests to account for it. See https://github.com/microsoft/TypeScript/pull/44624.
* Our `ngtsc` test setup calls the private `matchFiles` function from TS. I had to update our usage, because a new parameter was added.
* There was one place where we were setting the readonly `hasTrailingComma` property. I updated the usage to pass in the value when constructing the object instead.
* Some browser types were updated which meant that I had to resolve some trivial type errors.
* The downlevel decorators tranform was running into an issue where the Closure synthetic comments were being emitted twice. I've worked around it by recreating the class declaration node instead of cloning it.

crisbeto commented:
We talked through this with @alxhub and we decided to move forward with the PR since the TS 4.4 bug won't be a problem in g3. The only change I've made is to restore the support for TS 4.2.

AndrewKushnir commented:
[Presubmit](http://test/OCL:398361486:BASE:398361516:1632354400643:48c38529).

AndrewKushnir commented:
FYI presubmits (including a global one) are successful, so I'm removing the "presubmit" label.

alxhub reviewed (approved):
Reviewed-for: global-approvers

andrewseguin reviewed (approved):
LGTM for Elements

crisbeto commented:
They won't. There will be a follow-up PR to this to drop support once the other repos have upgraded.
